snuba/query/logical.py

```diff
@@ -63,6 +63,11 @@ def __init__(
             granularity=granularity,
         )
 
+    def __repr__(self) -> str:
+        from snuba.query.formatters.tracing import format_query
+
+        return "\n".join(format_query(self))
```

Comment:
This is the same code you have in the composite query. Can't the repr implementation be in the abstract query class entirely or is that a typing issue ?

Reply:
Yes it is a typing issue. The `DataSourceVisitor` and `JoinVisitor` may only visit `TSimpleDataSource`s, the abstract query is a `DataSource`

Reply:
I could maybe do it without those classes or change them but that seems like more work than re-defining the repr function. Open to suggestions